lib/phase/functional.js

```diff
@@ -0,0 +1,170 @@
+'use strict';
+
+const Joi = require('joi');
+const Hoek = require('hoek');
+const clone = require('clone');
+
+// Actual environment size is limited by space, not quantity.
+// We do this to force sd.yaml to be simpler.
+const MAX_ENVIRONMENT_VARS = 25;
+// If they are trying to execute >25 permutations,
+// maybe there is a better way to accomplish their goal.
+const MAX_PERMUTATIONS = 25;
+
+/**
+ * Generate the default list of jobs (normal order without main)
+ * @method getDefaultWorkflow
+ * @param  {Object}           jobs Object of jobName => jobDefinition
+ * @return {Array}                 List of jobs in order of definition (without main)
+ */
+function getDefaultWorkflow(jobs) {
+    const jobNames = Object.keys(jobs);
+
+    // Remove main
+    jobNames.splice(jobNames.indexOf('main'), 1);
+
+    return jobNames;
+}
+
+/**
+ * Make sure the Matrix they specified is valid
+ * @method validateJobMatrix
+ * @param  {Object}      job     Job to inspect
+ * @param  {String}      jobName Name of job (for error reporting)
+ * @return {Array}               List of errors
+ */
+function validateJobMatrix(job, jobName) {
+    let matrixSize = 1;
+    const errors = [];
+    const matrix = Hoek.reach(job, 'matrix', {
+        default: {}
+    });
+    const environment = Hoek.reach(job, 'environment', {
+        default: {}
+    });
+
+    if (Object.keys(matrix).length + Object.keys(environment).length > MAX_ENVIRONMENT_VARS) {
+        errors.push(`Job ${jobName}: "environment" and "matrix" can only have a combined ` +
+            ` maxiumum of ${MAX_ENVIRONMENT_VARS} environment variables defined`);
+    }
+
+    Object.keys(matrix).forEach((row) => {
+        matrixSize *= matrix[row].length;
+    });
+
+    if (matrixSize > MAX_PERMUTATIONS) {
+        errors.push(`Job ${jobName}: Matrix cannot contain >${MAX_PERMUTATIONS} permutations`);
```

Comment:
Can we say something like:

    `Job ${jobName}: Matrix cannot contain >${MAX_PERMUTATIONS} permutations. (Currently ${matrixSize})`

Same with environment above.